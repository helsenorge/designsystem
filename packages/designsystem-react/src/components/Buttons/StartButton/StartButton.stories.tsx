import React from 'react';
import {withKnobs, select, text} from '@storybook/addon-knobs';
import {storiesOf} from '@storybook/react';
import StartButton from './StartButton';
import {Icon} from '../../Icons';
import {allIcons} from '../../../../.storybook/knobs';

const stories = storiesOf('Buttons/StartButton', module);
stories.addDecorator(withKnobs);

stories.add('Default', () => (
  <StartButton variant={select('Variant', ['primary', 'secondary', 'tertiary'], 'primary')}>
    {text('Text', 'StartButton')}
  </StartButton>
));
