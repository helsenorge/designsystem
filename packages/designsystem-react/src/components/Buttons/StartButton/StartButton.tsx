import React, {useState} from 'react';
import {HTMLButtonProps, HTMLAnchorProps} from '../../../constants';
import Button, {ButtonVariants, ButtonSizes} from '../Button/Button';

type StartButtonVariants = 'primary' | 'secondary' | 'tertiary';
type StartButtonTag = 'a' | 'button';

interface StartButtonProps extends HTMLButtonProps, HTMLAnchorProps {
  variant?: StartButtonVariants;
  children: React.ReactNode;
  htmlTag?: StartButtonTag;
  loading?: boolean;
  onClick?: () => void;
}

const variantMap = {
  primary: 'fill',
  secondary: 'fill',
  tertiary: 'outline',
};

const sizeMap = {
  primary: 'large',
  secondary: 'medium',
  tertiary: 'medium',
};

const StartButton = React.forwardRef((props: StartButtonProps, ref: any) => {
  const {children, variant = 'primary', htmlTag = 'button', loading, ...rest} = props;
  return (
    <Button
      variant={variantMap[variant] as ButtonVariants}
      size={sizeMap[variant] as ButtonSizes}
      htmlTag={htmlTag}
      loading={loading}
      color="vein"
      {...rest}>
      {children}
    </Button>
  );
});

export default StartButton;
