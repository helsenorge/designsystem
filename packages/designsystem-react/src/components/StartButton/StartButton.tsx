import React from 'react';
import cx from 'classnames';
import {ButtonProps, Button} from '../Button';

interface StartButtonProps extends ButtonProps {}

// TODO: Move most of the logic out in a generic Button-component that the others inherit.
const StartButton = React.forwardRef((props: StartButtonProps, ref: any) => {
  const {children, variant, ...rest} = props;
  const classes = cx('is-start', {[`is-${variant}`]: variant === 'secondary' || variant === 'tertiary'});
  return (
    <Button className={classes} ref={ref} {...rest}>
      {children}
    </Button>
  );
});

export {StartButton};
