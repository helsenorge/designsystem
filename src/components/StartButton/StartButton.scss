@import '../../styling/helsenorge';

$start-button-primary-color-font: white;
$start-button-primary-color-background: $color-primary;

$start-button-secondary-color-font: white;
$start-button-secondary-color-background: $color-secondary;

$start-button-tertiary-color-font: black;
$start-button-tertiary-color-background: $color-tertiary;
$start-button-tertiary-color-border: $color-tertiary-relief;

.start-button {
  padding: 0 2.5rem;
  height: 5rem;
  border: 0.25rem solid transparent;
  &--primary {
    background-color: $start-button-primary-color-background;
    color: $start-button-primary-color-font;
  }
  &--secondary {
    background-color: $start-button-secondary-color-background;
    color: $start-button-secondary-color-font;
  }
  &--tertiary {
    background-color: $start-button-tertiary-color-background;
    color: $start-button-tertiary-color-font;
    border-color: $start-button-tertiary-color-border;
  }
}
