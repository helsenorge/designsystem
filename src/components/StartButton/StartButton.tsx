import React from 'react';
import cn from 'classnames';

// import './StartButton.scss';

// TODO: Make this inherit from a set of base variant types in constants.
export type StartButtonVariant = 'secondary' | 'tertiary' | string | undefined;

interface StartButtonProps
  extends React.DetailedHTMLProps<React.ButtonHTMLAttributes<HTMLButtonElement>, HTMLButtonElement> {
  children: React.ReactNode;
  variant?: StartButtonVariant;
  isLoading?: boolean;
}

// TODO: Move most of the logic out in a generic Button-component that the others inherit.
const StartButton = React.forwardRef((props: StartButtonProps, ref: any) => {
  const {children, variant, isLoading = false, ...rest} = props;
  const classes = cn(
    'button is-start',
    {[`is-${variant}`]: variant === 'secondary' || variant === 'tertiary'},
    {['is-loading']: isLoading},
  );
  return (
    // TODO: Add a 'as' prop so that the button can be either an 'a'-tag or 'button'-tag.
    <button className={classes} ref={ref} {...rest}>
      <span>{children}</span>
    </button>
  );
});

StartButton.defaultProps = {
  type: 'button',
};

export {StartButton};
