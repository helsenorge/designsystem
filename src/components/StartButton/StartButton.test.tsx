import React from 'react';
import {render} from '@testing-library/react';
import {StartButton} from './StartButton';

test('renders default variant correctly', (): void => {
  const {container} = render(<StartButton>StartButton</StartButton>);
  expect(container).toMatchSnapshot();
});
