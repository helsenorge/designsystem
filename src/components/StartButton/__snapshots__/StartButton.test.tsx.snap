// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders default variant correctly 1`] = `
<div>
  <button
    class="button is-start"
    type="button"
  >
    <span>
      StartButton
    </span>
  </button>
</div>
`;
